src/foam/box/CapabilityInterceptBox.js

```diff
@@ -0,0 +1,105 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.box',
+  name: "CapabilityInterceptReplyDecoratorBox",
+  extends: 'foam.box.ProxyBox',
+
+  documentation: `
+    This box decorates replyBox of the message with CapabilityInterceptBox
+    before sending the message to its delegate.
+  `,
+
+  methods: [
+    {
+      name: 'send',
+      javaCode: `
+        foam.core.X x = getX();
+
+        msg.getAttributes().put("replyBox",
+          new CapabilityInterceptBox.Builder(x)
+            .setDelegate((Box) msg.getAttributes().get("replyBox"))
+            .build());
+        getDelegate().send(msg);
+      `
+    }
+  ]
+});
+
+foam.CLASS({
+  package: 'foam.box',
+  name: 'CapabilityInterceptBox',
+  extends: 'foam.box.ProxyBox',
+
+  javaImports: [
+    'foam.box.RPCErrorMessage',
+    'foam.box.CapabilityRequiredRemoteException',
+    'foam.dao.ArraySink',
+    'foam.dao.DAO',
+    'foam.nanos.auth.AuthorizationException',
+    'foam.nanos.auth.User',
+    'foam.nanos.crunch.Capability',
+    'java.util.ArrayList',
+    'java.util.List',
+    'static foam.mlang.MLang.*'
+  ],
+
+  methods: [
+    {
+      name: 'send',
+      javaCode: `
+        foam.core.X x = getX();
+
+        FindCapabilityIntercept: {
```

Comment:
Is this necessary? Couldn't we just change everywhere that says

```Java
break FindCapabilityIntercept;
```

below to 

```Java
return super.send(msg);
```

? That would be preferable in my opinion.

Reply:
It would have to be two lines because Java doesn't allow returning void types for some reason. This would also make it easier to modify that behaviour later.

Would `super.send` be preferred over `getDelegate().send`?

Reply:
Edit: after changes related to your other comment there won't be as many repetitions of this, so the labelled block isn't useful anymore